user-mach4/lib/libcmucs/openp.3
.\" Copyright (c) 1990 Carnegie Mellon University
.\" All Rights Reserved.
.\" 
.\" Permission to use, copy, modify and distribute this software and its
.\" documentation is hereby granted, provided that both the copyright
.\" notice and this permission notice appear in all copies of the
.\" software, derivative works or modified versions, and any portions
.\" thereof, and that both notices appear in supporting documentation.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND CARNEGIE MELLON UNIVERSITY
.\" DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING ALL
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS.  IN NO EVENT
.\" SHALL CARNEGIE MELLON UNIVERSITY BE LIABLE FOR ANY SPECIAL, DIRECT,
.\" INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER
.\" RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF
.\" CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
.\" CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.\" Users of this software agree to return to Carnegie Mellon any
.\" improvements or extensions that they make and grant Carnegie the
.\" rights to redistribute these changes.
.\"
.\" Export of this software is permitted only after complying with the
.\" regulations of the U.S. Deptartment of Commerce relating to the
.\" Export of Technical Data.
.\"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.\" HISTORY
.\" $Log: openp.3,v $
# Revision 1.1  1996/03/01  14:51:54  sclawson
# Changes to install man pages.
#
.\" Revision 1.1.1.1  1995/05/04  06:56:39  sclawson
.\" New files.
.\"
.\" Revision 2.3  92/02/16  15:25:13  rpd
.\" 	Moved from libcs to libcmucs.
.\" 
.\" Revision 2.2  92/01/22  23:15:16  rpd
.\" 	Moved here from facilities's libcs.
.\" 	[92/01/19            rpd]
.\" 
.\" Revision 1.2  90/12/12  15:55:15  mja
.\" 	Add copyright/disclaimer for distribution.
.\" 
.\" 13-Nov-86  Andi Swimmer (andi) at Carnegie-Mellon University
.\" 	Revised for 4.3.
.\" 
.\" 24-Mar-86  Rudy Nedved (ern) at Carnegie-Mellon University
.\" 	Updated for 4.2BSD.
.\" 
.\" 05-Dec-79  Steven Shafer (sas) at Carnegie-Mellon University
.\" 	Created.
.\" 
.TH OPENP 3 3/24/86
.CM 4
.SH "NAME"
openp, fopenp \- search for file and open it
.SH "SYNOPSIS"
.B 
#include <sys/file.h>
.br
.B 
int openp (searchlist, path, buffer, flags, mode);
.br
.B 
char *searchlist, *path, *buffer;
.br
.B 
int flags, mode;

.B 
#include <sys/file.h>
.br
.B 
#include <stdio.h>
.br
.B 
FILE *fopenp (searchlist,path,buffer,type);
.br
.B 
char *searchlist, *path, *buffer ,*type;
.SH "DESCRIPTION"
.I 
Openp
and
.I 
fopenp
use
.IR searchp (3)
to search for a file, and open the file when and if it is
found.
The value returned will be the normal value of
.IR open (2)
or
.IR fopen (3s)
-- a file descriptor or FILE pointer on success, and a
\-1 or a 0 on failure.

.I 
Searchlist
is assumed to be a list of directory path names separated by colons;
one by one, these names are parsed and concatenated (with a
separating slash) onto
.I 
path
to form a complete file path name.
An attempt is then made to
open the file with this name; if successful,
.I 
openp
and
.I 
fopenp
return with a success indication; otherwise, searching
continues.
If success is achieved, then the resulting
pathname is copied into the string
.I 
buffer,
provided by the user.
The
.I 
open
and
.I 
fopen
calls are made with the appropriate
.I 
mode, flags
and
.I 
type
as specified by the parameter. The 
.IR open(2)
and
.IR fopen(3S)
define what
.I 
mode, flags
and
.I 
type
should be respectively.
.SH "SEE ALSO"
.IR searchp (3), 
.IR open (2), 
.IR fopen (3s), 
.IR wantread (3), 
.IR wantwrite (3),
.IR fwantread (3), 
.IR fwantwrite (3)
.SH "DIAGNOSTICS"
.I 
Openp
returns \-1 on error (no openable file found);
.I 
fopenp
returns 0.


gnu-osfmig/NEWS
-*- mode: outline -*- Time-stamp: "2003-08-25 20:13:52 jemarch"


* New in gnu-osfmig 0.01

** Exist as a separate package, out of the osfmk sources and building framework.

** All ODE stuff dissapears

** [autoconf]igurable


usr_random/src/usr/lib/libcmucs/openp.c
/*
 * Copyright (c) 1990 Carnegie Mellon University
 * All Rights Reserved.
 * 
 * Permission to use, copy, modify and distribute this software and its
 * documentation is hereby granted, provided that both the copyright
 * notice and this permission notice appear in all copies of the
 * software, derivative works or modified versions, and any portions
 * thereof, and that both notices appear in supporting documentation.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND CARNEGIE MELLON UNIVERSITY
 * DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING ALL
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS.  IN NO EVENT
 * SHALL CARNEGIE MELLON UNIVERSITY BE LIABLE FOR ANY SPECIAL, DIRECT,
 * INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER
 * RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF
 * CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
 * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *
 * Users of this software agree to return to Carnegie Mellon any
 * improvements or extensions that they make and grant Carnegie the
 * rights to redistribute these changes.
 *
 * Export of this software is permitted only after complying with the
 * regulations of the U.S. Deptartment of Commerce relating to the
 * Export of Technical Data.
 */
/*  openp, fopenp  --  search pathlist and open file
 *
 *  Usage:
 *	i = openp (path,file,complete,flags,mode)
 *	f = fopenp (path,file,complete,type)
 *	int i,flags,mode;
 *	FILE *f;
 *	char *path,*file,*complete,*type;
 *
 *  Openp searches for "file" in the pathlist "path";
 *  when the file is found and can be opened by open()
 *  with the specified "flags" and "mode", then the full filename
 *  is copied into "complete" and openp returns the file
 *  descriptor.  If no such file is found, openp returns -1.
 *  Fopenp performs the same function, using fopen() instead
 *  of open() and type instead of flags/mode; it returns 0 if no
 *  file is found.
 *
 *  HISTORY
 * $Log:	openp.c,v $
 * Revision 2.3  92/02/16  15:25:02  rpd
 * 	Moved from libcs to libcmucs.
 * 
 * Revision 2.2  92/01/22  23:14:46  rpd
 * 	Moved here from facilities's libcs.
 * 	[92/01/19            rpd]
 * 
 * Revision 1.2  90/12/11  17:57:28  mja
 * 	Add copyright/disclaimer for distribution.
 * 
 * 30-Apr-85  Steven Shafer (sas) at Carnegie-Mellon University
 *	Adapted for 4.2 BSD UNIX.  Added new parameter to openp.c;
 *	changed names of flags, mode, and type parameters to reflect
 *	current manual entries for open and fopen.
 *
 * 20-Nov-79  Steven Shafer (sas) at Carnegie-Mellon University
 *	Created for VAX.
 *
 */

#include <stdio.h>

int open();
int searchp();

static int flgs,mod,value;
static char *ftyp;
static FILE *fvalue;

static int func (fnam)
char *fnam;
{
	value = open (fnam,flgs,mod);
	return (value < 0);
}

static int ffunc (fnam)
char *fnam;
{
	fvalue = fopen (fnam,ftyp);
	return (fvalue == 0);
}

int openp (path,file,complete,flags,mode)
char *path, *file, *complete;
int flags, mode;
{
	register char *p;
	flgs = flags;
	mod = mode;
	if (searchp(path,file,complete,func) < 0)  return (-1);
	return (value);
}

FILE *fopenp (path,file,complete,ftype)
char *path, *file, *complete, *ftype;
{
	register char *p;
	ftyp = ftype;
	if (searchp(path,file,complete,ffunc) < 0)  return (0);
	return (fvalue);
}
